### PR TITLE
Use HierarchicalGeometryContainer to simplfiy CKF source link selection

### PR DESCRIPTION
**Pull request by XiaocongAi**
This PR simplify the CKF source link selection by using the hierarchical geometry container for the selection criteria.

### CODE REVIEW

**Comment by XiaocongAi** [comment]
Hi @msmk0 , could you help take a look at this PR to see if the `HierarchicalGeometryContainer` is used in the right way?

File: Core/include/Acts/TrackFinder/CKFSourceLinkSelector.hpp [hunk 1]
```diff
@@ -32,37 +51,32 @@ struct CKFSourceLinkSelector {
...
-        std::map<GeometryID::Value, VolumeNumMeasurements>;
-
-    // Global maximum chi2
-    double maxChi2 = std::numeric_limits<double>::max();
-
-    // Volume-level maximum chi2
-    VolumeChi2 volumeMaxChi2;
+    using Chi2Container =
+        Acts::HierarchicalGeometryContainer<GeometryElement<double>>;
+    using SourceLinkNumContainer =
```
[the remark this answers is not in the document]
Comment (XiaocongAi) [reply Core/include/Acts/TrackFinder/CKFSourceLinkSelector.hpp:64]: I see your point. But yeah, I expect they could be set at different hierarchy levels. So it's better to keep them separate.

File: Core/include/Acts/TrackFinder/CKFSourceLinkSelector.hpp [hunk 2]
```diff
@@ -108,70 +122,32 @@ struct CKFSourceLinkSelector {
...
+    // Find the cutoff value for the surface in the hierarchical geometry
+    // container. If not found, use global cutoff value
+    double chi2CutOff = std::numeric_limits<double>::max();
+    size_t numSourcelinksCutOff = std::numeric_limits<size_t>::max();
+    // -> Get the allowed maximum chi2 on this surface
+    auto chi2_it = m_config.chi2CutOffContainer.find(geoID);
+    if (chi2_it != m_config.chi2CutOffContainer.end()) {
+      chi2CutOff = chi2_it->value;
+    } else {
+      chi2CutOff = m_config.globalChi2CutOff;
```
[the remark this answers is not in the document]
Comment (XiaocongAi) [reply Core/include/Acts/TrackFinder/CKFSourceLinkSelector.hpp:138]: I think so. Is that something easy to implement in the container?

File: Core/include/Acts/TrackFinder/CKFSourceLinkSelector.hpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (msmk0) [reply Core/include/Acts/TrackFinder/CKFSourceLinkSelector.hpp:64]: Are these expected to be wildly different? If not I would still suggest to merge the two; you end up with less templates, less types, and less code. If it turns out to be a performance bottleneck later on, we can still switch back, but I strongly prefer to start with the simpler solution first.

File: Core/include/Acts/TrackFinder/CKFSourceLinkSelector.hpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (msmk0) [reply Core/include/Acts/TrackFinder/CKFSourceLinkSelector.hpp:138]: Yes, I think that should be relatively straight-forward. In that case, you would have to make sure during construction that the container always contains a global value.

I would not wait for the change in this PR though. Could you open a follow up issue to add support for this?

[the remark this answers is not in the document]
Comment (XiaocongAi) [reply Core/include/Acts/TrackFinder/CKFSourceLinkSelector.hpp:138]: I see. In that case, the global values must be explicitly provided. Agreed, let's do this in another PR.

File: Core/include/Acts/TrackFinder/CKFSourceLinkSelector.hpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (XiaocongAi) [reply Core/include/Acts/TrackFinder/CKFSourceLinkSelector.hpp:64]: OK, I don't have strong objection. Currently, we just use the global values for both of them.  Will be combined.

File: Tests/UnitTests/Core/TrackFinder/CombinatorialKalmanFilterTests.cpp [hunk 3]
```diff
@@ -289,20 +294,25 @@ BOOST_AUTO_TEST_CASE(comb_kalman_filter_zero_field) {
...
-  //-> pixel: 7
-  //-> strip: 8
-  sourcelinkSelectorConfig.volumeMaxChi2 = {{2, 7}, {3, 8}};
-  sourcelinkSelectorConfig.maxChi2 = 8;
-  // Set the allowed maximum number of source links to be large enough
-  sourcelinkSelectorConfig.maxNumSourcelinksOnSurface = 100;
+  // Chi2 cutoff at detector sensitive/layer/volume level implemented via
+  // hierarchical geometry container
+  sourcelinkSelectorConfig.chi2CutOffContainer =
+      Chi2Container(std::move(elements));
```
[the remark this answers is not in the document]
Comment (XiaocongAi) [reply Tests/UnitTests/Core/TrackFinder/CombinatorialKalmanFilterTests.cpp:311]: Unfortunately, it does not work. Let's add the additional constructor in another PR.

**Comment by XiaocongAi** [comment]
@msmk0.  I have combined the container for chi2 and nSourcelinks configuration. Could we merge it in and handle the other issues in a different PR?

**Comment by msmk0** [comment]
I approved the changes and the CI build runs through, but I can not merge. It looks like some checks are blocking it. Not sure what to do in that case. @paulgessinger Suggestions?

**Comment by paulgessinger** [comment]
Seems to have been a communication problem with the bots.

**Pull request merged**